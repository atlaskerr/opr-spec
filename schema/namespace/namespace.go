// Copyright Â© 2019 Titan Distribution Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package namespace

import (
	"github.com/atlaskerr/opr-spec/schema"

	"github.com/xeipuuv/gojsonschema"
)

var (
	createNamespaceFilepath = "namespace/create-namespace.schema.json"
)

// CreateNamespaceSchema returns a gojsonschema.Schema for validation.
func CreateNamespaceSchema() *gojsonschema.Schema {
	path := "namespace/create-namespace.schema.json"
	mediatype := schema.MediaTypeCreateNamespace
	return schema.Load(path, mediatype)
}

// BatchCreateNamespaceSchema returns a gojsonschema.Schema for validation.
func BatchCreateNamespaceSchema() *gojsonschema.Schema {
	path := "namespace/batch-create-namespace.schema.json"
	mediatype := schema.MediaTypeBatchCreateNamespace
	return schema.Load(path, mediatype)
}

// DeleteNamespaceSchema returns a gojsonschema.Schema for validation.
func DeleteNamespaceSchema() *gojsonschema.Schema {
	path := "namespace/delete-namespace.schema.json"
	mediatype := schema.MediaTypeDeleteNamespace
	return schema.Load(path, mediatype)
}

// BatchDeleteNamespaceSchema returns a gojsonschema.Schema for validation.
func BatchDeleteNamespaceSchema() *gojsonschema.Schema {
	path := "namespace/batch-delete-namespace.schema.json"
	mediatype := schema.MediaTypeBatchDeleteNamespace
	return schema.Load(path, mediatype)
}
