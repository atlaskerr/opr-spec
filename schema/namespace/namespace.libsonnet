// Copyright Â© 2019 Titan Distribution Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

local openapi = 'openapi';
local jsonschema = 'jsonschema';
local V7 = 'http://json-schema.org/draft-07/schema#';

local jid = {
  namespace: 'https://titan-registry.io/schema/namespace',
  createNamespace: 'https://titan-registry.io/schema/namespace/namespace-create',
  listNamespaces: 'https://titan-registry.io/schema/namespace-list',
};

local uint64 = {
  type: 'integer',
  minimum: 0,
  maximum: 18446744073709551615,
};

local mapStringString(output=jsonschema) = {
  type: 'object',
  [if output == jsonschema
  then 'patternProperties'
  else if output == openapi
  then 'x-patternProperties']: {
    '.{1,}': {
      type: 'string',
    },
  },
};

local namespace(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.namespace,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    name: { type: 'string' },
    storageLimit: uint64,
    storageUsed: uint64,
    repoLimit: uint64,
    repoCount: uint64,
    labels: mapStringString(output),
    status: {
      type: 'string',
      enum: [
        'ACTIVE',
        'TERMINATING',
      ],
    },
  },
};

local createNamespace(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.createNamespace,
  [if output == jsonschema then '$schema']: V7,
  title: 'Create Namespace',
  type: 'object',
  properties: {
    storageLimit: uint64,
    repoLimit: uint64,
    labels: mapStringString(output),
  },
};

local listNamespaces(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.listNamespaces,
  [if output == jsonschema then '$schema']: V7,
  title: 'List Namespaces',
  type: 'object',
  properties: {
    namespaces: {
      type: 'array',
      items: namespace(''),
    },
  },
};

{
  namespace:: namespace,
  createNamespace:: createNamespace,
  listNamespaces:: listNamespaces,
}
