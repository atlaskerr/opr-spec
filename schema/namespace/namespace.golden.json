{
   "labels": {
      "costCenter": "cs-foo",
      "manager": "Tom Ripen",
      "team": "team-foo"
   },
   "name": "namespace-foo",
   "repoCount": 23,
   "repoLimit": 100,
   "storageLimit": 548298572358,
   "storageUsed": 580324752
}
